File: 2021_random_set_self_practice/using_python/자물쇠와열쇠.py
def reverse_board(board):
    R = len(board)
    C = len(board[0])
    return [[board[i][j] ^ 1 for j in range(C)] for i in range(R)]


def rotate(board):
    return [col[::-1] for col in zip(*board)]


def cpy_board(board):
    return [row[:] for row in board]


def extract(board, sy, sx, length):
    y1, x1 = sy, sx
    y2, x2 = sy + length, sx + length
    new_board = [board[y][x1:x2] for y in range(y1, y2)]
    return new_board


def is_equal(board1, board2, cnt_hole):
    R = len(board1)
    C = len(board1[0])

    cnt = 0
    for i in range(R):
        for j in range(C):
            if board2[i][j] == '_':
                continue
            if board1[i][j] != board2[i][j]:
                return False
            if board1[i][j] == 1:
                cnt += 1
    if cnt == cnt_hole:
        return True
    return False


def is_matched(key_rotated, new_lock, cnt_hole):
    move_y, move_x = len(new_lock) // 3, len(new_lock[0]) // 3
    for sy in range(1, 2 * move_y):
        for sx in range(1, 2 * move_x):
            new_board = extract(new_lock, sy, sx, len(new_lock) // 3)
            if is_equal(key_rotated, new_board, cnt_hole):
                return True
    return False


def solution(key, lock):
    lock = reverse_board(lock)
    length = len(lock)
    new_lock = [['_' for _ in range(3 * length)] for _ in range(3 * length)]
    cnt_hole = 0
    for i in range(len(lock)):
        for j in range(len(lock[0])):
            new_lock[i + length][j + length] = lock[i][j]
            if lock[i][j] == 1:
                cnt_hole += 1

    key_rotated = cpy_board(key)
    for _ in range(4):
        if is_matched(key_rotated, new_lock, cnt_hole):
            return True
        key_rotated = rotate(key_rotated)
    return False


if __name__ == "__main__":
    print(solution([[0, 0, 0], [1, 0, 0], [0, 1, 1]], [[1, 1, 1], [1, 1, 0], [1, 0, 1]]))

File: 2021_random_set_self_practice/using_java/SrcCodes/src/자물쇠와열쇠/자물쇠와열쇠.java
package 자물쇠와열쇠;


class Match {
    int[][] key;
    int[][] lock;
    int cntHole;
    int lockLength;
    int keyLength;

    // init
    public Match(int[][] key, int[][] lock) {
        this.key = key;
        this.keyLength = key.length;

        this.lockLength = lock.length;
        this.lock = new int[3 * this.lockLength][3 * this.lockLength];
        // init this.lock (3배 확대)
        this.cntHole = 0;
        for (int i = 0; i < 3 * this.lockLength; i++) {
            for (int j = 0; j < 3 * this.lockLength; j++) {
                this.lock[i][j] = -1;
            }
        }
        for (int i = 0; i < lock.length; i++) {
            for (int j = 0; j < lock.length; j++) {
                this.lock[i + this.lockLength][j + this.lockLength] = lock[i][j] ^ 1;
                if (lock[i][j] == 0) {  // hole 있는 부분
                    this.cntHole++;
                }
            }
        }
    }

    public boolean go() {
        int N = this.lockLength + this.keyLength;
        int move = this.lockLength - this.keyLength;

        int[][] keyRotated = this.key;
        for (int k = 0; k < 4; k++) {
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    int sy = i + move;
                    int sx = j + move;
                    int[][] box = extract(sy, sx);
                    if (doMatch(box, keyRotated)) {
                        return true;
                    }
                }
            }
            keyRotated = this.rotate(keyRotated);
        }
        return false;
    }

    public int[][] extract(int sy, int sx) {
        int[][] ret = new int[this.keyLength][this.keyLength];
        for (int i = 0; i < this.keyLength; i++) {
            for (int j = 0; j < this.keyLength; j++) {
                ret[i][j] = this.lock[i + sy][j + sx];
            }
        }
        return ret;
    }

    public boolean doMatch(int[][] box, int[][] keyRotated) {
        int cnt = 0;
        for (int i = 0; i < this.keyLength; i++) {
            for (int j = 0; j < this.keyLength; j++) {
                if (box[i][j] == -1) {
                    continue;
                }

                if (box[i][j] != keyRotated[i][j]) {
                    return false;
                }
                if (box[i][j] == 1) {
                    cnt++;
                }
            }
        }

        return cnt == this.cntHole;
    }

    public int[][] rotate(int[][] org) {
        int N = org.length;
        int[][] ret = new int[N][N];
        for (int col = 0; col < N; col++) {
            for (int row = 0; row < N; row++) {
                ret[col][N - row - 1] = org[row][col];
            }
        }
        return ret;
    }
}


class Solution {
    public boolean solution(int[][] key, int[][] lock) {
        Match match = new Match(key, lock);
        return match.go();
    }

    public static void main(String[] args) {
        int[][] key = {{0, 0, 0}, {1, 0, 0}, {0, 1, 1}};
        int[][] lock = {{1, 1, 1}, {1, 1, 0}, {1, 0, 1}};
        Solution sol = new Solution();
        System.out.println(sol.solution(key, lock));
    }
}